### DISCUSSION TITLE
How to solve FileNotFoundError problem？

### DISCUSSION THREAD

**Discussion by forestbat** [pyg-team/pytorch_geometric#5744] [Q&A]
When I run test code in [this article](https://pytorch-geometric.readthedocs.io/en/latest/notes/introduction.html#learning-methods-on-graphs)，it occurs mistake like this：
`FileNotFoundError: Could not find module 'C:\Python310\Lib\site-packages\torch_sparse\_convert_cuda.pyd' (or one of its dependencies). Try using the full path with constructor syntax.`
And I tried to install wheel from [this link](https://pytorch-geometric.com/whl/torch-1.12.1+cu116.html), however the main function still crashed with FileNotFoundError.
I'm sure there is _convert_cuda.pyd in torch_sparse directory.
And when I run `python -c "import torch; print(torch.__version__)"`
the output is: `1.12.1+cpu`

Is there any connection with torch version? How to solve my problem?

**Comment by rusty1s** [comment]
How did you install `torch-sparse`? Does `import torch_sparse` work for you?

**Reply by forestbat** [reply]
At first I install `torch-sparse` with `pip install torch-scatter torch-sparse torch-cluster torch-spline-conv torch-geometric -f https://data.pyg.org/whl/torch-1.12.0+cu116.html`, and when it occurs wrong, I tried`pip install C:\……\torch_sparse-0.6.15-cp310-cp310-win_amd64.whl`，however it still has no effect.

`import torch` doesn't work for me.

**Reply by forestbat** [reply]
> How did you install `torch-sparse`? Does `import torch_sparse` work for you?

And before I ask here, I have run `pip3 install torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/cu116`.

**Reply by rusty1s** [reply]
So did you download the wheel from https://data.pyg.org/whl/torch-1.12.0%2Bcu116/torch_sparse-0.6.15-cp310-cp310-win_amd64.whl directly? If you unpack, you can see that the necessary `*.pyd` files are indeed included there. My assumption is that you may have some broken version currently installed. Can you try to uninstall all versions first before installing again?

**Reply by forestbat** [reply]
Yes，I have downloaded the wheel from that link directly，and even if I replaced `_convert_cuda.pyd` with file from the wheel，it still has wrong.
I reinstalled `torch-scatter torch-sparse torch-cluster torch-spline-conv torch-geometric` ，however idea still tells me `FileNotFoundError`.
The md5 of my `_convert_cuda.pyd` is 3010e3c0601e7370b0d159677148b8da，please check if there is something wrong if you can.

**Reply by rusty1s** [reply]
So the file `C:\Python310\Lib\site-packages\torch_sparse\_convert_cuda.pyd'` actually exists but it still reports a `FileNotFoundError`? This is super weird :( What happens if you try to install the CPU version?

**Reply by forestbat** [reply]
It looks like has no effect:

```
PS C:\Users\UserName\IdeaProjects\station-simulate> pip install torch-scatter torch-sparse torch-cluster torch-spline-conv torch-geometric -f https://data.pyg.org/whl/torch-1.12.0+cpu.html
Looking in links: https://data.pyg.org/whl/torch-1.12.0+cpu.html
Requirement already satisfied: torch-scatter in c:\python310\lib\site-packages (2.0.9)
Requirement already satisfied: torch-sparse in c:\python310\lib\site-packages (0.6.15)
Requirement already satisfied: torch-cluster in c:\python310\lib\site-packages (1.6.0)
Requirement already satisfied: torch-spline-conv in c:\python310\lib\site-packages (1.2.1)
Requirement already satisfied: torch-geometric in c:\python310\lib\site-packages (2.1.0.post1)
Requirement already satisfied: scipy in c:\python310\lib\site-packages (from torch-sparse) (1.8.1)
Requirement already satisfied: scikit-learn in c:\python310\lib\site-packages (from torch-geometric) (1.1.2)
Requirement already satisfied: pyparsing in c:\python310\lib\site-packages (from torch-geometric) (3.0.9)
Requirement already satisfied: tqdm in c:\python310\lib\site-packages (from torch-geometric) (4.64.1)
Requirement already satisfied: numpy in c:\python310\lib\site-packages (from torch-geometric) (1.22.4)
Requirement already satisfied: requests in c:\python310\lib\site-packages (from torch-geometric) (2.28.1)
Requirement already satisfied: jinja2 in c:\python310\lib\site-packages (from torch-geometric) (3.1.2)
Requirement already satisfied: MarkupSafe>=2.0 in c:\python310\lib\site-packages (from jinja2->torch-geometric) (2.1.1)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in c:\python310\lib\site-packages (from requests->torch-geometric) (1.26.12)
Requirement already satisfied: certifi>=2017.4.17 in c:\python310\lib\site-packages (from requests->torch-geometric) (2022.9.14)
Requirement already satisfied: idna<4,>=2.5 in c:\python310\lib\site-packages (from requests->torch-geometric) (3.4)
Requirement already satisfied: charset-normalizer<3,>=2 in c:\python310\lib\site-packages (from requests->torch-geometric) (2.1.1)
Requirement already satisfied: threadpoolctl>=2.0.0 in c:\python310\lib\site-packages (from scikit-learn->torch-geometric) (3.1.0)
Requirement already satisfied: joblib>=1.0.0 in c:\python310\lib\site-packages (from scikit-learn->torch-geometric) (1.2.0)
Requirement already satisfied: colorama in c:\python310\lib\site-packages (from tqdm->torch-geometric) (0.4.5)
```
If you can't answer this question, I will put the discussion to bug report issues.

**Reply by rusty1s** [reply]
Keep in mind that you always need to uninstall first (at best run this multiple times):
```
pip uninstall torch-scatter torch-sparse torch-cluster torch-spline-conv
```

**Reply by forestbat** [reply]
Now I use pyg's cpu version，it tells me this：
```
C:\Python310\python.exe C:\Users\UserName\IdeaProjects\station-simulate\cora-test.py 
C:\Python310\lib\site-packages\torch_geometric\datasets\qm9.py:21: UserWarning: Failed to initialize NumPy: module compiled against API version 0x10 but this version of numpy is 0xf (Triggered internally at  ..\torch\csrc\utils\tensor_numpy.cpp:68.)
  conversion = torch.tensor([
Downloading https://github.com/kimiyoung/planetoid/raw/master/data/ind.cora.x
Downloading https://github.com/kimiyoung/planetoid/raw/master/data/ind.cora.tx
Downloading https://github.com/kimiyoung/planetoid/raw/master/data/ind.cora.allx
Downloading https://github.com/kimiyoung/planetoid/raw/master/data/ind.cora.y
Downloading https://github.com/kimiyoung/planetoid/raw/master/data/ind.cora.ty
Downloading https://github.com/kimiyoung/planetoid/raw/master/data/ind.cora.ally
Downloading https://github.com/kimiyoung/planetoid/raw/master/data/ind.cora.graph
Downloading https://github.com/kimiyoung/planetoid/raw/master/data/ind.cora.test.index
Processing...
Traceback (most recent call last):
  File "C:\Users\UserName\IdeaProjects\station-simulate\cora-test.py", line 6, in <module>
    dataset = Planetoid(root='/tmp/Cora', name='Cora')
  File "C:\Python310\lib\site-packages\torch_geometric\datasets\planetoid.py", line 93, in __init__
    super().__init__(root, transform, pre_transform)
  File "C:\Python310\lib\site-packages\torch_geometric\data\in_memory_dataset.py", line 50, in __init__
    super().__init__(root, transform, pre_transform, pre_filter)
  File "C:\Python310\lib\site-packages\torch_geometric\data\dataset.py", line 87, in __init__
    self._process()
  File "C:\Python310\lib\site-packages\torch_geometric\data\dataset.py", line 170, in _process
    self.process()
  File "C:\Python310\lib\site-packages\torch_geometric\datasets\planetoid.py", line 151, in process
    data = read_planetoid_data(self.raw_dir, self.name)
  File "C:\Python310\lib\site-packages\torch_geometric\io\planetoid.py", line 22, in read_planetoid_data
    items = [read_file(folder, prefix, name) for name in names]
  File "C:\Python310\lib\site-packages\torch_geometric\io\planetoid.py", line 22, in <listcomp>
    items = [read_file(folder, prefix, name) for name in names]
  File "C:\Python310\lib\site-packages\torch_geometric\io\planetoid.py", line 107, in read_file
    out = torch.Tensor(out)
ValueError: too many dimensions 'matrix'
```

**Reply by rusty1s** [reply]
Never seen this error before. Can you debug the content of `out` before line 107?

**Reply by forestbat** [reply]
exception: (<class 'ValueError'>, ValueError("too many dimensions 'matrix'"), <traceback object at 0x000002803459D880>)
![图片](https://user-images.githubusercontent.com/23413915/196913939-b05a54be-2667-4b23-bba0-3a56b1c87626.png)
I don‘t know how to get debug content with txt，so I can only give you a picture.

**Reply by rusty1s** [reply]
Mh, this is weird. Looks to be an issue with `pickle`. What is the version of `pickle` you are using?

**Reply by forestbat** [reply]
```
python -c "import pickle; print(pickle.format_version)"
4.0
```
